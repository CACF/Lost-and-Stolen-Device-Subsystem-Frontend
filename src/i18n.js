/*
Copyright (c) 2018 Qualcomm Technologies, Inc.
All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted (subject to the limitations in the disclaimer below) provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
* Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
* Neither the name of Qualcomm Technologies, Inc. nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE GRANTED BY THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

import i18n from 'i18next';
import LanguageDetector from 'i18next-browser-languagedetector';
 
i18n
  .use(LanguageDetector)
  .init({
    // we init with resources
    resources: {
      en: {
        translations: {
          "welcomeApp": "Welcome to LSMS",
          "homeLink"  : "Home",
          "dashboardLink": "Dashboard",
          "newCaseLink": "New Case",
          "viewCaseLink": "View Case",
          "updateCaseLink": "Update Case",
          "searchCaseLink": "Search Cases",
          "allCasesLink"  : "All Cases",
          "pendingCaseLink": "Pending Cases",
          "blockedCaseLink": 'Blocked Cases',
          "recoveredCaseLink": 'Recovered Cases',
          "CaseStatusLink": "Case Status",
        }
      },
      es: {
        translations: {
          "welcomeApp": "Bienvenido a LSMS",
          "homeLink"  : "Casa",
          "dashboardLink": "Tablero",
          "newCaseLink"  : "Nuevo caso",
          "viewCaseLink": "Ver caso",
          "updateCaseLink": "Actualizar caso",
          "searchCaseLink"  : "Casos de b√∫squeda",
          "allCasesLink"  : "Todos los casos",
          "pendingCaseLink": "Casos pendientes",
          "blockedCaseLink": 'Casos bloqueados',
          "recoveredCaseLink": 'Casos recuperados',
          "CaseStatusLink": "Estado del caso",
        }
      },
      id: {
        translations: {
          "welcomeApp": "Selamat datang di LSMS",
          "homeLink"  : "Rumah",
          "dashboardLink": "Dasbor",
          "newCaseLink"  : "Kasus baru",
          "viewCaseLink": "Lihat Kasus",
          "updateCaseLink": "Perbarui Kasus",
          "searchCaseLink"  : "Kasus Pencarian",
          "allCasesLink"  : "Semua Kasus",
          "pendingCaseLink": "Kasus yang Tertunda",
          "blockedCaseLink": 'Kasus yang Diblokir',
          "recoveredCaseLink": 'Kasus yang Dipulihkan',
          "CaseStatusLink": "Status Kasus",
        }
      }
    },
    fallbackLng: 'en',
    debug: true,
 
    // have a common namespace used around the full app
    ns: ['translations'],
    defaultNS: 'translations',
 
    keySeparator: false, // we use content as keys
 
    interpolation: {
      escapeValue: false, // not needed for react!!
      formatSeparator: ','
    },
 
    react: {
      wait: true
    }
  });
 
export default i18n;