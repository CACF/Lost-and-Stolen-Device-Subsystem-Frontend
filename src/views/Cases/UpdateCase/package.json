{
  "name": "UpdateCase",
  "version": "0.0.0",
  "private": true,
  "main": "./UpdateCase.js"
}
