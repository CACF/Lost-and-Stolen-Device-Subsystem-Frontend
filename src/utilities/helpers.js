/*
SPDX-License-Identifier: BSD-4-Clause-Clear
Copyright (c) 2018-2019 Qualcomm Technologies, Inc.
All rights reserved.
Redistribution and use in source and binary forms, with or without
modification, are permitted (subject to the limitations in the disclaimer
below) provided that the following conditions are met:

   - Redistributions of source code must retain the above copyright notice,
   this list of conditions and the following disclaimer.
   - Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.
   - All advertising materials mentioning features or use of this software,
   or any deployment of this software, or documentation accompanying any
   distribution of this software, must display the trademark/logo as per the
   details provided here:
   https://www.qualcomm.com/documents/dirbs-logo-and-brand-guidelines
   - Neither the name of Qualcomm Technologies, Inc. nor the names of its
   contributors may be used to endorse or promote products derived from this
   software without specific prior written permission.


SPDX-License-Identifier: ZLIB-ACKNOWLEDGEMENT
Copyright (c) 2018-2019 Qualcomm Technologies, Inc.
This software is provided 'as-is', without any express or implied warranty.
In no event will the authors be held liable for any damages arising from
the use of this software.
Permission is granted to anyone to use this software for any purpose,
including commercial applications, and to alter it and redistribute it
freely, subject to the following restrictions:

   - The origin of this software must not be misrepresented; you must not
   claim that you wrote the original software. If you use this software in a
   product, an acknowledgment is required by displaying the trademark/logo as
   per the details provided here:
   https://www.qualcomm.com/documents/dirbs-logo-and-brand-guidelines
   - Altered source versions must be plainly marked as such, and must not
   be misrepresented as being the original software.
   - This notice may not be removed or altered from any source distribution.

NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE GRANTED BY
THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT
NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER
OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

import React from 'react';
import axios from 'axios';
import Base64 from 'base-64';
import Swal from 'sweetalert2';
import withReactContent from 'sweetalert2-react-content';
import i18n from './../i18n'
import {
  ENGLISH_REGEX,
  SPANISH_REGEX,
  INDONESIAN_REGEX,
  ENGLISH_FULL_NAME,
  SPANISH_FULL_NAME,
  INDONESIAN_FULL_NAME
} from './constants';
import settings from '../settings';

const MySwal = withReactContent(Swal);
const { defaultLanguage } = settings.appDetails;

/**
 * Create API, Keycloak and Tyk URLs based on settings.json file
 */
const {host: apiHost, port: apiPort, version: apiVersion, use: apiUse} = settings.api;
const {host: apimanHost, port: apimanPort, clientId: apimanClientId, use: apimanUse} = settings.apiman;
const {appName} = settings.appDetails;

let BASE_URL = '';

if(apiUse) {
 BASE_URL = `${apiHost}${apiPort ? ':'+ apiPort: ''}${apiVersion}`;
} else if(apimanUse) {
 BASE_URL = `${apimanHost}${apimanPort ? ':'+ apimanPort: ''}${apimanClientId}`;
}

/**
 * generate axios instance for calling http requests with common URL.
 *
 * @type {AxiosInstance}
 */
export const instance = axios.create({
    baseURL: BASE_URL, // QA Build
});

/**
 * This function return an header for http request with Keycloak token
 *
 * @returns {{Authorization: string, "Content-Type": string}}
 */
export function getAuthHeader (token) {
  let accessToken = localStorage.getItem('token');
  if(token) {
    accessToken = token;
  }
  return {
    'Authorization': 'Bearer ' + accessToken,
    'Content-Type': 'application/json'
  }
}

export function range(start, limit, step) {

  if (arguments.length === 1) {
    limit = start;
    start = 0;
  }

  limit = limit || 0;
  step = step || 1;

  for (var result = []; (limit - start) * step > 0; start += step) {
    result.push(start);
  }
  return result;
}

/**
 * This function return the encode user info stored in local storage from Keycloak
 *
 * @returns {any}
 */
export function getUserInfo () {
  return JSON.parse(Base64.decode(localStorage.getItem('userInfo')))
}

/**
 * this function get all groups of currently loggedIn user
 *
 * @param resources
 * @returns {string}
 */
export function getUserGroups (resources) {
  let groups = '';
  let groupDetails = (((resources || {}).realm_access || {}).roles || []);
  // Remove default group first
  let index = groupDetails.indexOf('uma_authorization');
  if (index !== -1)
      groupDetails.splice(index, 1);
  if(groupDetails.length > 0) {
    groups = groupDetails;
  }
  return groups;
}

/**
 * This functions get users' groups assigned by Keycloak and see if user has access to this application
 *
 * @param groups
 * @returns {boolean}
 */
export function isPage401(groups) {
  let pageStatus = false; // assume it's not page401
  pageStatus = (groups.length > 0) ? false : true; // if groups exist then that's not page401
  if (! pageStatus) { // if groups exist then we apply another check
    pageStatus = groups.find(app => app.split('_')[0] === appName) ? false : true; // app name is same as role assigned
  }
  return pageStatus;
}

export function SweetAlert(params){
  let title = params.title
  let message = params.message
  let type = params.type

  MySwal.fire({
    title: <p>{title}</p>,
    text: message,
    type: type,
    confirmButtonText: i18n.t('button.ok')
  })
}

/**
 * Generic Errors handling for Axios
 *
 * @param context
 * @param error
 */
export function errors (context, error) {
  if(typeof error !== 'undefined'){
    if(typeof error.response === 'undefined') {
      SweetAlert({
        title: i18n.t('error'),
        message: i18n.t('serverNotResponding'),
        type:'error'
      })
      //toast.error('API Server is not responding or You are having some network issues');
    } else {
      if(error.response.status === 400) {
        SweetAlert({
          title: i18n.t('error'),
          message: error.response.data.message,
          type:'error'
        })
        //toast.error(error.response.data.message);
      } else if(error.response.status === 401) {
        SweetAlert({
          title: i18n.t('error'),
          message: i18n.t('sessionExpired'),
          type:'error'
        })
        //toast.error('Your session has been expired, please wait');
        setTimeout(() => {
          window.location.reload();
        }, 3000);
      } else if(error.response.status === 403) {
        SweetAlert({
          title: i18n.t('error'),
          message: i18n.t('credentialMatch'),
          type:'error'
        })
        //toast.error('These credential do not match our records.');
      } else if(error.response.status === 404) {
        SweetAlert({
          title: i18n.t('error'),
          message: error.response.data.message,
          type:'error'
        })
        //toast.error(error.response.data.message);
      } else if(error.response.status === 405) {
        SweetAlert({
          title: i18n.t('error'),
          message: i18n.t('wrongHttp'),
          type:'error'
        })
        //toast.error('You have used a wrong HTTP verb');
      } else if(error.response.status === 406) {
        SweetAlert({
          title: i18n.t('error'),
          message: error.response.data.message,
          type:'error'
        })
        //toast.error(error.response.data.message);
      } else if(error.response.status === 409) {
        SweetAlert({
          title: i18n.t('error'),
          message: error.response.data.message,
          type:'error'
        })
        //toast.error(error.response.data.message, { autoClose: 10000 });
      } else if(error.response.status === 422) {
        let errors = error.response.data.messages;
        for (var key in errors) {
           var caseErrors = errors[key];
           for(var i in caseErrors) {
             SweetAlert({
               title: i18n.t('error'),
               message: caseErrors[i][0],
               type:'error'
             });
              //toast.error(caseErrors[i][0], { autoClose: 10000 });
           }
        }
      } else if(error.response.status >= 500) {
        SweetAlert({
          title: i18n.t('error'),
          message: i18n.t('serverNotResponding'),
          type:'error'
        })
        //toast.error("API Server is not responding or You are having some network issues", { autoClose: 8000 });
      }
    }
  }
}

export function languageCheck(text) {
  if(ENGLISH_REGEX.test(text) && defaultLanguage==="en"){
    return true;
  }else if(SPANISH_REGEX.test(text) && defaultLanguage==="es"){
    return true;
  }else return INDONESIAN_REGEX.test(text) && defaultLanguage === "id";
}

export function fullNameCheck(text){
  if(ENGLISH_FULL_NAME.test(text) && defaultLanguage==="en"){
    return true;
  }else if(SPANISH_FULL_NAME.test(text) && defaultLanguage==="es"){
    return true;
  }else return INDONESIAN_FULL_NAME.test(text) && defaultLanguage === "id";
}