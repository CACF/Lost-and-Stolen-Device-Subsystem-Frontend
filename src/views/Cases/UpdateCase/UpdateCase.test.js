import React from 'react';
import {shallow, mount, render} from 'enzyme';
import UpdateCase from './UpdateCase';
import {I18nextProvider} from "react-i18next";
import i18n from "./../../../i18nTest"
import sinon from 'sinon'
import mockAxios from 'jest-mock-axios';
import {BrowserRouter as Router} from 'react-router-dom';
import {getUserInfo} from "../../../utilities/helpers";

const mockKcProps = {
  'isTokenExpired': sinon.spy()
}

const match = { params: { tracking_id: 'IUJ34ET5' } }

const mockHeader = {
  "headers": {
    "Authorization": "Bearer null",
    "Content-Type": "application/json"
  }
}

const historyMock = { push: jest.fn() };

let responseObj = {
  data: {
        comments: [{
          comment: "Testing comment",
          comment_date: "2018-10-29T10:58:31.189731",
          user_id: "9a4403c2-5a48-4b79-9f30-02c0cc7799e0",
          username: "test"
        }],
        creator: {
          username: 'test',
          user_id: '9a4403c2-5a48-4b79-9f30-02c0cc7799e0'
        },
        device_details: {
          imeis: ['123456787655345'],
          description: 'test description',
          brand: 'test brand',
          model_name: 'test model',
          msisdns: ['03231234567']
        },
        incident_details: {
          incident_date: '2018-01-01',
          incident_nature: 'Test'
        },
        personal_details: {
          full_name: 'test',
          email: 'test@example.com',
          dob: null,
          gin: null,
          address: 'test address',
          number: null
        },
        status: 'Recovered',
        tracking_id: 'IUJ34ET5',
        updated_at: '2018-01-01 10:10:10',
        created_at: '2018-10-29 10:34:34'
      }
}

afterEach(() => {
  // cleaning up the mess left behind the previous test
  mockAxios.reset();
});

describe('UpdateCase component', () => {

  beforeEach(() => {
    // values stored in tests will also be available in other tests unless you run
    localStorage.clear();
    localStorage.userInfo = 'eyJzdWIiOiI5YTQ0MDNjMi01YTQ4LTRiNzktOWYzMC0wMmMwY2M3Nzk5ZTAiLCJwcmVmZXJyZWRfdXNlcm5hbWUiOiJxdWFsY29tbSJ9';
  })

  test("if renders correctly", () => {
    const wrapper = shallow(<UpdateCase match={match}/>);
    expect(wrapper).toMatchSnapshot()
  });

  test('If UpdateCase exists',()=>{
    const wrapper = shallow(<UpdateCase match={match}/>);
    expect(wrapper.exists()).toBe(true);
  })

  test('UpdateCase have certain states', () => {
    const wrapper = mount(
      <I18nextProvider i18n={i18n}>
        <UpdateCase handleCaseStatus={() => {
          }} kc={mockKcProps} match={match}/>
      </I18nextProvider>
    );
    expect(wrapper.find('UpdateCase').state().loading).toBe(true)
    expect(wrapper.find('UpdateCase').state().data).toEqual(null)
  })

  test('if validations work fine', () => {
    let responseInvalidObj = {
      data: {
            comments: [{
              comment: "Testing comment",
              comment_date: "2018-10-29T10:58:31.189731",
              user_id: "9a4403c2-5a48-4b79-9f30-02c0cc7799e0",
              username: "test"
            }],
            creator: {
              username: 'test',
              user_id: '9a4403c2-5a48-4b79-9f30-02c0cc7799e0'
            },
            device_details: {
              imeis: ['123456787655345'],
              description: 'test description',
              brand: 'test brand',
              model_name: 'test model',
              msisdns: ['03231234567']
            },
            incident_details: {
              incident_date: '2018-01-01',
              incident_nature: 'Test'
            },
            personal_details: { // invalid mock personal details
              full_name: null,
              email: null,
              dob: null,
              gin: null,
              address: null,
              number: null
            },
            status: 'Recovered',
            tracking_id: 'IUJ34ET5',
            updated_at: '2018-01-01 10:10:10',
            created_at: '2018-10-29 10:34:34'
          }
    }
    const wrapper = mount(
        <Router>
          <I18nextProvider i18n={i18n}>
            <UpdateCase handleCaseStatus={() => {
              }} kc={mockKcProps} match={match}/>
          </I18nextProvider>
        </Router>
      );
    mockAxios.mockResponse(responseInvalidObj)
    wrapper.update()
    expect(mockAxios.get).toHaveBeenCalledWith('/case/IUJ34ET5', mockHeader);
    wrapper.find('form').simulate('submit');
    expect(wrapper.find('Formik').state().errors.full_name).toEqual('This field is Required');
    expect(wrapper.find('Formik').state().errors.oneOfFields).toEqual('One of the fields is required');
  });

  test('if DOB is from future', () => {
    let responseInvalidObj = {
      data: {
            comments: [{
              comment: "Testing comment",
              comment_date: "2018-10-29T10:58:31.189731",
              user_id: "9a4403c2-5a48-4b79-9f30-02c0cc7799e0",
              username: "test"
            }],
            creator: {
              username: 'test',
              user_id: '9a4403c2-5a48-4b79-9f30-02c0cc7799e0'
            },
            device_details: {
              imeis: ['123456787655345'],
              description: 'test description',
              brand: 'test brand',
              model_name: 'test model',
              msisdns: ['03231234567']
            },
            incident_details: {
              incident_date: '2018-01-01',
              incident_nature: 'Test'
            },
            personal_details: {
              full_name: null,
              email: null,
              dob: '2020-12-27', // invalid mock DOB
              gin: null,
              address: null,
              number: null
            },
            status: 'Recovered',
            tracking_id: 'IUJ34ET5',
            updated_at: '2018-01-01 10:10:10',
            created_at: '2018-10-29 10:34:34'
          }
    }
    const wrapper = mount(
        <Router>
          <I18nextProvider i18n={i18n}>
            <UpdateCase handleCaseStatus={() => {
              }} kc={mockKcProps} match={match}/>
          </I18nextProvider>
        </Router>
      );
    mockAxios.mockResponse(responseInvalidObj)
    wrapper.update()
    expect(mockAxios.get).toHaveBeenCalledWith('/case/IUJ34ET5', mockHeader);
    wrapper.find('form').simulate('submit');
    expect(wrapper.find('Formik').state().errors.dob).toEqual("Date of Birth can't be in future");
  });

  test('if DOB is from past', () => {
    let resDOBInvalidObj = {
      data: {
            comments: [{
              comment: "Testing comment",
              comment_date: "2018-10-29T10:58:31.189731",
              user_id: "9a4403c2-5a48-4b79-9f30-02c0cc7799e0",
              username: "test"
            }],
            creator: {
              username: 'test',
              user_id: '9a4403c2-5a48-4b79-9f30-02c0cc7799e0'
            },
            device_details: {
              imeis: ['123456787655345'],
              description: 'test description',
              brand: 'test brand',
              model_name: 'test model',
              msisdns: ['03231234567']
            },
            incident_details: {
              incident_date: '2018-01-01',
              incident_nature: 'Test'
            },
            personal_details: {
              full_name: null,
              email: null,
              dob: '1880-12-27', // invalid mock DOB
              gin: null,
              address: null,
              number: null
            },
            status: 'Recovered',
            tracking_id: 'IUJ34ET5',
            updated_at: '2018-01-01 10:10:10',
            created_at: '2018-10-29 10:34:34'
          }
    }
    const wrapper = mount(
        <Router>
          <I18nextProvider i18n={i18n}>
            <UpdateCase handleCaseStatus={() => {
              }} kc={mockKcProps} match={match}/>
          </I18nextProvider>
        </Router>
      );
    mockAxios.mockResponse(resDOBInvalidObj)
    wrapper.update()
    expect(mockAxios.get).toHaveBeenCalledWith('/case/IUJ34ET5', mockHeader);
    wrapper.find('form').simulate('submit');
    expect(wrapper.find('Formik').state().errors.dob).toEqual("Date of Birth can't be that old");
  });

  test('if email is invalid', () => {
    let resInvalidEmailObj = {
      data: {
            creator: {
              username: 'test',
              user_id: '9a4403c2-5a48-4b79-9f30-02c0cc7799e0'
            },
            device_details: {
              imeis: ['123456787655345'],
              description: 'test description',
              brand: 'test brand',
              model_name: 'test model',
              msisdns: ['03231234567']
            },
            incident_details: {
              incident_date: '2018-01-01',
              incident_nature: 'Test'
            },
            personal_details: {
              full_name: null,
              email: 'asdfasfsdf', // invalid mock Email
              dob: null,
              gin: null,
              address: null,
              number: null
            },
            status: 'Recovered',
            tracking_id: 'IUJ34ET5',
            updated_at: '2018-01-01 10:10:10',
            created_at: '2018-10-29 10:34:34'
          }
    }
    const wrapper = mount(
        <Router>
          <I18nextProvider i18n={i18n}>
            <UpdateCase handleCaseStatus={() => {
              }} kc={mockKcProps} match={match}/>
          </I18nextProvider>
        </Router>
      );
    mockAxios.mockResponse(resInvalidEmailObj)
    wrapper.update()
    expect(mockAxios.get).toHaveBeenCalledWith('/case/IUJ34ET5', mockHeader);
    wrapper.find('form').simulate('submit');
    expect(wrapper.find('Formik').state().errors.email).toEqual("Invalid email address");
  });

  test('if comment text is greater than 1000 characters', () => {
    let resInvalidCommentObj = {
      data: {
            comments: [{
              comment: "test comment",
              comment_date: "2018-10-29T10:58:31.189731",
              user_id: "9a4403c2-5a48-4b79-9f30-02c0cc7799e0",
              username: "test"
            }],
            creator: {
              username: 'test',
              user_id: '9a4403c2-5a48-4b79-9f30-02c0cc7799e0'
            },
            device_details: {
              imeis: ['123456787655345'],
              description: 'test description',
              brand: 'test brand',
              model_name: 'test model',
              msisdns: ['03231234567']
            },
            incident_details: {
              incident_date: '2018-01-01',
              incident_nature: 'Test'
            },
            personal_details: {
              full_name: null,
              email: 'asdfasfsdf',
              dob: null,
              gin: null,
              address: null,
              number: null
            },
            status: 'Recovered',
            tracking_id: 'IUJ34ET5',
            updated_at: '2018-01-01 10:10:10',
            created_at: '2018-10-29 10:34:34'
          }
    }
    const wrapper = mount(
        <Router>
          <I18nextProvider i18n={i18n}>
            <UpdateCase handleCaseStatus={() => {
              }} kc={mockKcProps} match={match}/>
          </I18nextProvider>
        </Router>
      );
    mockAxios.mockResponse(resInvalidCommentObj)
    wrapper.update()
    expect(mockAxios.get).toHaveBeenCalledWith('/case/IUJ34ET5', mockHeader);
    const inputtedData = {
      case_comment: {
        target: {
          value: 'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate eget, arcu. In enim justo, rhoncus ut, imperdiet a, venenatis vitae, justo. Nullam dictum felis eu pede mollis pretium. Integer tincidunt. Cras dapibus. Vivamus elementum semper nisi. Aenean vulputate eleifend tellus. Aenean leo ligula, porttitor eu, consequat vitae, eleifend ac, enim. Aliquam lorem ante, dapibus in, viverra quis, feugiat a, tellus. Phasellus viverra nulla ut metus varius laoreet. Quisque rutrum. Aenean imperdiet. Etiam ultricies nisi vel augue. Curabitur ullamcorper ultricies nisi. Nam eget dui. Etiam rhoncus. Maecenas tempus, tellus eget condimentum rhoncus, sem quam semper libero, sit amet adipiscing sem neque sed ipsum. Na',
          name: 'case_comment'
        }
      }
    };
    wrapper.find('textarea').at(0).simulate('change', inputtedData.case_comment);
    wrapper.update();
    wrapper.find('form').simulate('submit');
    expect(wrapper.find('Formik').state().errors.case_comment).toEqual("Must be 1000 characters or less");
  });

  test('if componentDidMount render values correctly', () => {
    const wrapper = mount(
        <Router>
          <I18nextProvider i18n={i18n}>
            <UpdateCase handleCaseStatus={() => {
              }} kc={mockKcProps} match={match}/>
          </I18nextProvider>
        </Router>
      );
    mockAxios.mockResponse(responseObj)
    wrapper.update()
    expect(mockAxios.get).toHaveBeenCalledWith('/case/IUJ34ET5', mockHeader);
    expect(wrapper.find('UpdateCase').state().loading).toBe(false);
    expect(wrapper.find('UpdateCase').state().caseTrackingId).toEqual(responseObj.data.tracking_id);
    expect(wrapper.find('.view-box').find('table')).toHaveLength(4);
    const inputtedData = {
      case_comment: {
        target: {
          value: 'testing comment',
          name: 'case_comment'
        }
      },
      full_name: {
        target: {
          value: 'test-name',
          name: 'full_name'
        }
      },
      dob: {
        target: {
          value: '1990-01-01',
          name: 'dob'
        }
      },
      gin: {
        target: {
          value: '12345678912345',
          name: 'gin'
        }
      },
      address: {
        target: {
          value: 'test-address',
          name: 'address'
        }
      },
      alternate_number: {
        target: {
          value: '03001234567',
          name: 'alternate_number'
        }
      },
      email: {
        target: {
          value: 'test@example.com',
          name: 'alternate_number'
        }
      }
    };
    const submittedCase = {
      status_args: {
        user_id : '9a4403c2-5a48-4b79-9f30-02c0cc7799e0',
        username : 'qualcomm',
        case_comment: 'testing comment'
      },
      personal_details: {
        address: "test-address",
        dob: "1990-01-01",
        email: "test@example.com",
        full_name: "test-name",
        gin: "12345678912345",
        number: "03001234567"
      }
    }
    const updateResponse = {
      data: {"tracking_id": "IUJ34ET5", "message": "Case updated successfully"}
    }
    wrapper.find('input').at(0).simulate('change', inputtedData.full_name);
    wrapper.find('input').at(1).simulate('change', inputtedData.dob);
    wrapper.find('input').at(2).simulate('change', inputtedData.gin);
    wrapper.find('input').at(3).simulate('change', inputtedData.address);
    wrapper.find('input').at(4).simulate('change', inputtedData.email);
    wrapper.find('input').at(5).simulate('change', inputtedData.alternate_number);
    wrapper.find('textarea').at(0).simulate('change', inputtedData.case_comment);
    wrapper.update();
    const submitButton = wrapper.find('button').at(1);
    submitButton.simulate('submit');
    expect(mockAxios.put).toHaveBeenCalledWith('/case/IUJ34ET5', submittedCase, mockHeader);
    mockAxios.mockResponse(updateResponse)
    wrapper.update()
    expect(wrapper.find('UpdateCase').state().caseSubmitted).toBe(true);
  })
});