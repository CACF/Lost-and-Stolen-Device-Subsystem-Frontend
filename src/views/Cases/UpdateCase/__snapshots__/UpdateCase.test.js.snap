// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateCase component if renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Translate(UpdateCase)
    match={
      Object {
        "params": Object {
          "tracking_id": "IUJ34ET5",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "bindI18n": "languageChanged loaded",
      "bindStore": "added removed",
      "children": [Function],
      "i18n": undefined,
      "match": Object {
        "params": Object {
          "tracking_id": "IUJ34ET5",
        },
      },
      "ns": Array [
        "translations",
      ],
      "nsMode": "default",
      "omitBoundRerender": true,
      "translateFuncName": "t",
      "usePureComponent": false,
      "wait": false,
      "withRef": false,
    },
    "ref": null,
    "rendered": [Function],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "bindI18n": "languageChanged loaded",
        "bindStore": "added removed",
        "children": [Function],
        "i18n": undefined,
        "match": Object {
          "params": Object {
            "tracking_id": "IUJ34ET5",
          },
        },
        "ns": Array [
          "translations",
        ],
        "nsMode": "default",
        "omitBoundRerender": true,
        "translateFuncName": "t",
        "usePureComponent": false,
        "wait": false,
        "withRef": false,
      },
      "ref": null,
      "rendered": [Function],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
